std/stdio.d

```diff
@@ -2859,55 +2859,25 @@ struct LockingTextReader
         return _front;
     }
 
-    /* Read a utf8 sequence from the file, removing the chars from the stream.
-    Returns an empty result when at EOF. */
-    private char[] takeFront(return ref char[4] buf)
-    {
-        import std.utf : stride, UTFException;
-        {
-            immutable int c = FGETC(cast(_iobuf*) _f._p.handle);
-            if (c == EOF)
-                return buf[0 .. 0];
-            buf[0] = cast(char) c;
-        }
-        immutable seqLen = stride(buf[]);
-        foreach(ref u; buf[1 .. seqLen])
-        {
-            immutable int c = FGETC(cast(_iobuf*) _f._p.handle);
-            if (c == EOF) // incomplete sequence
-                throw new UTFException("Invalid UTF-8 sequence");
-            u = cast(char) c;
-        }
-        return buf[0 .. seqLen];
-    }
-
-    /* Read a utf8 sequence from the file into _front, putting the chars back so
-    that they can be read again.
+    /* Read a char from the file into _front, putting the char back so that it
+    can be read again.
     Destroys/closes the file when at EOF. */
     private void readFront()
     {
         import std.exception : enforce;
         import std.utf : decodeFront;
 
-        char[4] buf;
-        auto chars = takeFront(buf);
+        immutable int c = FGETC(cast(_iobuf*) _f._p.handle);
 
-        if (chars.empty)
+        if (c == EOF)
         {
             .destroy(_f);
             assert(empty);
             return;
         }
 
-        auto s = chars;
-        _front = decodeFront(s);
-
-        // Put everything back.
-        foreach(immutable i; 0 .. chars.length)
-        {
-            immutable c = chars[$ - 1 - i];
-            enforce(ungetc(c, cast(FILE*) _f._p.handle) == c);
-        }
+        _front = cast(char) c;
+        enforce(ungetc(c, cast(FILE*) _f._p.handle) == c);
```

Comment:
https://github.com/D-Programming-Language/phobos/pull/3698 only calls ungetc in the destructor.